pkg/smokescreen/conntrack/conn_tracker.go

```
@@ -15,19 +17,92 @@ type Tracker struct {
+	CalculatedAt    time.Time
+	ConnSuccessRate float64
+	TotalConns      int
+}
+
+// StartNewConnSuccessRateTracker creates a new ConnSuccessRateTracker with a specific calculation interval at which
+// ConnSuccessRateStats will be recomputed, and a time window to calculate those statistics over.
+func StartNewConnSuccessRateTracker(calculationInterval time.Duration, calculationWindow time.Duration) *ConnSuccessRateTracker {
+	newSuccessTracker := &ConnSuccessRateTracker{
+		ConnAttempts: cache.New(calculationWindow, time.Second*10),
+	}
+	newSuccessTracker.ConnSuccessRateStats.Store(ConnSuccessRateStats{CalculatedAt: time.Now(), ConnSuccessRate: 100, TotalConns: 0})
+
+	go func() {
+		for {
+			var total, succeeded int
+			for _, success := range newSuccessTracker.ConnAttempts.Items() {
+				total++
+				if success.Object.(bool) {
+					succeeded++
+				}
+			}
+			var successRate float64
+			// Avoid divide by zero errors
+			if total == 0 {
+				successRate = float64(100)
+			} else {
+				successRate = (float64(succeeded) / float64(total)) * 100
+			}
+			newSuccessTracker.ConnSuccessRateStats.Store(ConnSuccessRateStats{CalculatedAt: time.Now(), ConnSuccessRate: successRate, TotalConns: total})
+
+			time.Sleep(calculationInterval)
+		}
+	}()
+
+	return newSuccessTracker
+}
+
+func NewTracker(idle time.Duration, statsc statsd.ClientInterface, logger *logrus.Logger, sd atomic.Value, successRateTracker *ConnSuccessRateTracker) *Tracker {
 	return &Tracker{
-		Map:          &sync.Map{},
-		ShuttingDown: sd,
-		Wg:           &sync.WaitGroup{},
-		IdleTimeout:  idle,
-		statsc:       statsc,
+		Map:                &sync.Map{},
+		ShuttingDown:       sd,
+		Wg:                 &sync.WaitGroup{},
+		IdleTimeout:        idle,
+		statsc:             statsc,
+		SuccessRateTracker: successRateTracker,
+	}
+}
+
+// RecordAttempt stores the result of the most recent connection attempt for a destination.
+func (tr *Tracker) RecordAttempt(dest string, success bool) error {
```

You might consider giving this function a void return type (i.e., remove `error`) and just quietly returning if `tr.SuccessRateTracker == nil`. That might simplify the code. It looks like you're not doing anything interesting with the error anyway in `recordConnectionAttempt()`.

Since this (and `ReportConnectionSuccessRate()`) seems to be testing for a misconfiguration, would you be able to come up with a way to surface misconfiguration earlier (instead of logging potentially a lot of errors)?

---

Sounds good to me! I don't see a great way to detect a misconfiguration before either of these functions are called, since there's no way to tell at startup time whether the caller intends to access these stats. Failing silently instead sgtm!